File: src/leetcode/Q3/Solution.js
var lengthOfLongestSubstring = function(s) {
    var map = {};
    var longest = 0;
    var pointer = 0, flag = 0;
    var N = s.length;
    while (flag <= N - 1) {
        var c = s[flag];
        var pos = -1;
        if (map.hasOwnProperty(c)) {
            pos = map[c];
        }
        if (pos < pointer) {
            map[c] = flag;
            var length = flag - pointer + 1;
            if (length > longest) {
                longest = length;
            }
        } else {
            pointer = pos + 1;
            map[c] = flag;
        }
        flag++;
    }
    return longest;
};

File: src/leetcode/java/Q3/Solution.java
package Q3;

import java.util.HashMap;

/**
 * Created by hao on 7/22/16.
 */
public class Solution {
    public int lengthOfLongestSubstring(String s) {
        HashMap<Character, Integer> map = new HashMap<>();
        int pointer = 0, flag = 0;
        int N = s.length();
        int longest = 0;
        while (flag <= N - 1) {
            Character c = s.charAt(flag);
            Integer pos = map.get(c);
            if (pos == null) pos = -1;
            if (pos < pointer) {
                map.put(c, flag);
                int current = flag - pointer + 1;
                if (current > longest) {
                    longest = current;
                }
            } else {
                pointer = pos + 1;
                map.put(c, flag);
            }
            flag++;
        }
        return longest;
    }
}